internal/authctx/selfmanaged.go

```diff
@@ -0,0 +1,232 @@
+/*
+Copyright © 2023 VMware, Inc. All Rights Reserved.
+SPDX-License-Identifier: MPL-2.0
+*/
+
+package authctx
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/pkg/errors"
+	oidcapi "go.pinniped.dev/generated/latest/apis/supervisor/oidc"
+	"go.pinniped.dev/pkg/oidcclient/pkce"
+	"go.pinniped.dev/pkg/oidcclient/state"
+	"golang.org/x/oauth2"
+)
+
+const (
+	tokenEndpointSuffix         = "oauth2/token"
+	authorizationEndpointSuffix = "oauth2/authorize"
+	redirectURL                 = "http://127.0.0.1/callback"
+	pinnipedCLIClientID         = oidcapi.ClientIDPinnipedCLI
+	contextTimeout              = 60 * time.Second
+	extraIDToken                = "id_token"
+	federationDomainPath        = "provider/pinniped"
+
+	mdKeyAuthToken    = "Authorization"
+	authTokenPrefix   = "Bearer "
+	mdKeyAuthIDToken  = "grpc-metadata-x-user-id"
+	mdKeyRefreshToken = "grpc-metadata-x-refresh-token"
+)
+
+type smSession struct {
+	sharedOauthConfig             *oauth2.Config
+	issuerURL, username, password string
+
+	pkceCodePair pkce.Code
+	stateVal     state.State
+}
+
+// todo: proxy support is not added for the self-managed flow. Add it when there is a requirement.
+func getSMUserAuthCtx(pinnipedURL, uName, password string) (metadata map[string]string, err error) {
+	if pinnipedURL == "" || uName == "" || password == "" {
+		return nil, errors.New("Invalid auth configuration for self_managed")
+	}
+
+	session, err := initSession(pinnipedURL, uName, password)
+	if err != nil {
+		return nil, err
+	}
+
+	expectedRedirectURL, err := url.Parse(session.sharedOauthConfig.RedirectURL)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse expected redirect URL %s", session.sharedOauthConfig.RedirectURL)
+	}
+
+	actualRedirectURL, err := session.initiateAuthorizeRequestUnamePwd()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to initiate authorize request with issuer %s", session.issuerURL)
+	}
+
+	// Check that the redirect was to the expected location.
+	if actualRedirectURL.Scheme != expectedRedirectURL.Scheme ||
```

Comment:
Too many conditions, can we not reverse and use this in the logic?

Reply:
what are you suggesting to reverse here?